### PR TITLE
#18: Shelve/Unshelve

### PR DESCRIPTION
**Pull request by incidentist**
This adds the following commands:
* Shelve
* Unshelve
* Unshelve and Keep
* Abort Unshelve
* Continue Unshelve

### CODE REVIEW

**Comment by incidentist** [comment]
@hdpoliveira What do you think?

File: src/commands.ts [hunk 1]
```diff
@@ -872,6 +872,28 @@ export class CommandCenter {
 		}
 	}
 
+	@command('hg.shelve', { repository: true })
+	public async shelve(repository: Repository) {
+		var options: ShelveOptions = {}
+		var shelveName = await interaction.inputShelveName();
```
Comment (hdpoliveira) [inline src/commands.ts:878]: In mercurial the user can specify a name as well as a description for the shelve. Maybe after getting the name we ask for an optional description (this may be done in a different PR)

File: src/commands.ts [hunk 2]
```diff
@@ -872,6 +872,28 @@ export class CommandCenter {
 		}
 	}
 
+	@command('hg.shelve', { repository: true })
+	public async shelve(repository: Repository) {
+		var options: ShelveOptions = {}
```
Comment (hdpoliveira) [inline src/commands.ts:877]: Use const when possible

File: src/repository.ts [hunk 3]
```diff
@@ -958,6 +961,26 @@ export class Repository implements IDisposable {
         return true;
     }
 
+    async shelve(options: ShelveOptions) {
+        return await this.run(Operation.Shelve, () => this.repository.shelve(options));
+    }
+
+    async unshelve(options: UnshelveOptions) {
+        return await this.run(Operation.Shelve, () => this.repository.unshelve(options));
```
Comment (hdpoliveira) [inline src/repository.ts:969]: I'm not very familiar with the Operation thing, but don't we need one for Unshelve?

File: src/interaction.ts [hunk 4]
```diff
@@ -191,14 +191,31 @@ export namespace interaction {
     export function warnNoRollback(this: void) {
         return window.showWarningMessage(localize('no rollback', "Nothing to rollback to."));
     }
-
+
```
Comment (hdpoliveira) [inline src/interaction.ts:194]: This looks like it doesn't belong.

We should automate code formatting to avoid these in the future.

File: src/commands.ts (hunk 1 above)
Comment (incidentist) [reply src/commands.ts:878]: [sigh] So, the shelve extension hasn't been touched in 7 years and isn't well-suited to tooling. Unlike other hg commands, there's no way to get parsable output from `hg shelve --list`. The format is `<shelve name><some number of spaces>(<human readable timestamp)<some number of spaces> changes to: <commit message of the change the shelve is based off>\n<description>`. So, no machine-readable timestamp, no commit hash, no consistent field separator. Any parser or regex would be confused by parentheses or linebreaks in the shelve name or description. It turns out TortoiseHG gets around this by bypassing the shelve extension completely and storing patches in its own .hg subfolder! This PR uses an [undocumented `-q` flag](https://www.mercurial-scm.org/repo/hg/file/tip/mercurial/shelve.py#l654) (😬) to just get the list of shelve names. It could be thrown off by linebreaks in the shelve name, but that's a pathological case.

One option is to have the extension parse shelve metadata directly. The file format is straightforward, but obviously the format could be changed from underneath us. If it ever happens, it'll be in a separate PR.

File: src/repository.ts (hunk 3 above)
Comment (incidentist) [reply src/repository.ts:969]: I don't really understand it either. The Git extension uses the same Operation for stash and unstash, so I'm doing the same.

**Comment by incidentist** [comment]
I created #95 last night to track progress on automated testing. I actually tried getting it working before working on this PR so that I could test these new commands easily, but couldn't get it working properly.

File: src/commands.ts (hunk 1 above)
Comment (hdpoliveira) [reply src/commands.ts:878]: -q is not undocumented, it appears when you write "hg shelve --help --verbose" :wink:
